feat(AIP-203): require field behavior

toumorokoshi opened the pull request:
To reduce the likelihood of missed field behavior annotations,
require it for all message feilds.

further rationale is explained in the change of AIP-203.

toumorokoshi committed:
feat(AIP-203): require field behavior

To reduce the likelihood of missed field behavior annotations,
require it for all message feilds.

further rationale is explained in the change of AIP-203.

bgrant0607 commented on aip/general/0203.md:
```diff
@@ -16,18 +16,19 @@ behavior (for example, to optimize client library signatures).
-above. However, it is never mandatory to explicitly describe a field as
-optional.
+APIs **must** use the `google.api.field_behavior` annotation on every message
+field and have one of [`REQUIRED`][], [`IMMUTABLE`][], or [`OUTPUT_ONLY`][].
```
Isn't immutable orthogonal to required, optional, and output-only?

Also, I know we don't like input_only, but what about input_only? Have we developed an alternative for sensitive fields?

toumorokoshi replied:
Sorry! I meant optional. 

I don't think we can require anything on `INPUT_ONLY` and `OUTPUT_ONLY` besides not making them conflict, since the implicit default is bi-directional.

bgrant0607 replied:
Don't REQUIRED and OPTIONAL imply they are input rather than output?

toumorokoshi replied:
Yes: I was trying to think of a way to ensure we have meaningful field behavior.
Hence I was thinking you need one of REQUIRED / OPTIONAL / OUTPUT_ONLY, and you
shouldn't have more than one of the above simultaneously.

toumorokoshi replied:
updated to state that all relevant annotations must be added.

toumorokoshi committed:
addressing feedback.

- fix IMMUTABLE to OPTIONAL among required attributes.

bgrant0607 commented on aip/general/0203.md:
```diff
@@ -162,11 +163,36 @@ the user's behalf.
+
+## History
+
+In 2023-05, field behavior was made mandatory due to required and immutable
```
Immutable is no longer mentioned above. I agree it makes sense to require it in the case that an attribute is immutable (i.e., a runtime error will be returned in the case that a client attempts to update the value) and is expected to remain immutable forever.

toumorokoshi replied:
Should I add something to clarify further? I can go modify the immutable section
to clarify the backwards-compatibility of each behavior in the relevant
annotation section.

toumorokoshi replied:
Actually I think the wording has to be updated quite a bit on immutable (it
refers specifically to requests), if it's okay I'll leave it to a different PR.

bgrant0607 replied:
That's fine, but I wouldn't mention immutable here, then.

toumorokoshi replied:
Okay I think I understand what you're saying. I've updated the wording above to
state that *all* relevant annotations must be used. I've included immutable as
an example of a field_behavior that, if added or removed, can result in a
backwards-incompatible change. Although it's arguable that any field_behavior
change can result in a backwards-incompatible client change.

toumorokoshi replied:
okay updated wording, PTAL.

toumorokoshi committed:
addressing feedback.

clarifying that *all* field behavior annotations are required.

toumorokoshi committed:
Merge branch 'master' into toum/require-field-behavior

bgrant0607 reviewed (approved):
Thanks. LGTM. We can clarify more in the future if necessary.

toumorokoshi committed:
Merge branch 'master' into toum/require-field-behavior

toumorokoshi committed:
Merge branch 'master' into toum/require-field-behavior

toumorokoshi committed:
Merge branch 'master' into toum/require-field-behavior

shwoodard commented on aip/general/0203.md:
```diff
@@ -16,18 +16,19 @@ behavior (for example, to optimize client library signatures).
-well-understood field behavior, such as a field's being required, immutable, or
-output only:
+APIs use the `google.api.field_behavior` annotation to describe well-understood
+field behavior, such as a field's being required or immutable.
```
`s/field's/field/`

> ...such as a field being required or immutable.

is, I think, grammatically correct and more readable. I know this is how it is in the existing version, but the possessive tense is wrong here, I think.

toumorokoshi replied:
agreed and good call, fixed.

shwoodard commented on aip/general/0203.md:
```diff
@@ -16,18 +16,19 @@ behavior (for example, to optimize client library signatures).
-above. However, it is never mandatory to explicitly describe a field as
-optional.
+APIs **must** use the `google.api.field_behavior` annotation on every message
+field and must include every relevant option in this AIP.
```
"must" should be bold, `**must**`

toumorokoshi replied:
ack, fixed.

shwoodard commented on aip/general/0203.md:
```diff
@@ -16,18 +16,19 @@ behavior (for example, to optimize client library signatures).
-above. However, it is never mandatory to explicitly describe a field as
-optional.
+APIs **must** use the `google.api.field_behavior` annotation on every message
+field and must include every relevant option in this AIP.
```
I don't follow this second phrase. Do you mean,

> ...every available option

IOW, are you trying to say,

> APIs **must** include the `google.api.field_behavior` annotation on every message field and the value **must** be amongst the relevant options in this AIP.

?

toumorokoshi replied:
I do'nt think we have to be worried about the option being among the set in this
AIP: it's an enum after all.

I was trying to say that *if* the option describes the behavior of the field
(e.g. Optional + Input Only), then both must be applied.

I'll try to find some better wording here.

toumorokoshi replied:
Ok, added some better wording, PTAL.

Hupdy replied:
A

shwoodard commented on aip/general/0203.md:
```diff
@@ -16,18 +16,19 @@ behavior (for example, to optimize client library signatures).
+field and must include every relevant option in this AIP.
+
+Fields with no annotation are interpreted as `OPTIONAL` for
+backwards-compatility, but this annotation **must not** be omitted.
```
nit: I think it would have a *little* more punch as follows:

> for backwards-compatibility. Nonetheless, this annotation...

toumorokoshi replied:
It feels the same to me, but I'm not opinionated. fixed.

shwoodard commented on aip/general/0203.md:
```diff
@@ -162,11 +163,37 @@ the user's behalf.
+### Requiring field behavior
+
+The field behavior annotation explicitly states the intended behavior of the
+resource which enables better documentation. In addition, requiring the
```
> of the resource

?

isn't of the field?

Maybe it would be good to say,

> By including the field behavior annotation for each field, the overall behavior that the resource exhibits is more clearly defined.

?

toumorokoshi replied:
Used that wording, and tried to clarify a little bit more.

shwoodard commented on aip/general/0203.md:
```diff
@@ -162,11 +163,37 @@ the user's behalf.
+
+The field behavior annotation explicitly states the intended behavior of the
+resource which enables better documentation. In addition, requiring the
+annotation forces the API author to explicitly consider the behavior during
```
nit: s/during/while/

shwoodard commented on aip/general/0203.md:
```diff
@@ -162,11 +163,37 @@ the user's behalf.
+Modifying field behavior after initial authoring results in
+backwards-incompatible changes in clients. For example, making an optional field
+required results in backwards-incompatible changes in the method signature of
+an RPC, or the require property for an [IaC][] client.
```
Please normalize line length.

shwoodard commented on aip/general/0203.md:
```diff
@@ -162,11 +163,37 @@ the user's behalf.
+annotation forces the API author to explicitly consider the behavior during
+authoring of the API.
+
+Modifying field behavior after initial authoring results in
```
"initially authoring" or "initial authorship"

shwoodard commented on aip/general/0203.md:
```diff
@@ -162,11 +163,37 @@ the user's behalf.
+
+Modifying field behavior after initial authoring results in
+backwards-incompatible changes in clients. For example, making an optional field
+required results in backwards-incompatible changes in the method signature of
```
I think a comma is needed after "required".

shwoodard commented on aip/general/0203.md:
```diff
@@ -162,11 +163,37 @@ the user's behalf.
+Modifying field behavior after initial authoring results in
+backwards-incompatible changes in clients. For example, making an optional field
+required results in backwards-incompatible changes in the method signature of
+an RPC, or the require property for an [IaC][] client.
```
"or the require property for an [IAC][] client"--what do you mean? I'm not sure I get this clause of the sentence.

toumorokoshi replied:
I was trying to say that the resource in an IaC client will change in a
backwards-incompatible way - but "require property" was the wrong way to say it,
probably.

I tried to reword, PTAL.

shwoodard commented on aip/general/0203.md:
```diff
@@ -162,11 +163,37 @@ the user's behalf.
+In 2023-05, field behavior was made mandatory due to field behavior which define
+client behavior (e.g. required, output_only, immutable) often being missed,
+resulting in poor quality clients that had to deal with backward-incompatible
+changes with a major version revision.
```
I think this paragraph would be easier to read if broken into multiple sentences. Also, "with a major version" should be "within a major version", I think. Consider the following:

> In 2023-05 `field_behavior` was made mandatory. Prior to this change, the annotation was often omitted. Its values, e.g. `REQUIRED`, `OUTPUT_ONLY`, and `IMMUTABLE`, are relied upon to produce high quality clients. Further, when the value is added after the fact or changes, within a major version, it is backwards-incompatible, which is likely to break clients.

toumorokoshi replied:
ack, used your suggestion.

shwoodard commented on aip/general/0203.md:
```diff
@@ -162,11 +163,37 @@ the user's behalf.
+changes with a major version revision.
+
+The cost of clients and API users addressing breaking changes outweighed the
+cost of a one-line annotation per field during API authoring.
```
Having this as a separate paragraph makes it hard to see that it relates to the `field_behavior`, discussed in the previous paragraph. TBH, I don't think you need this sentence at all, but if you would like it, I would append the following to the previous suggestion,

> The benefits of requiring `field_behavior`, at the time that the API is authored, surpass the costs to clients and API users of not doing so.

If you decide to keep your sentence as is, I think "outweighed" needs to change to "outweigh" or "are outweighed" to prevent the "sentence" from being a sentence fragment.

toumorokoshi replied:
ack. I think it is important to clarify that the cost to clients is actually
greater than the cost to API authors: I've experience anecdotally that the cost
to the API owner is (erroneously) weighted much more heavily than long-term
client costs.

I'll take your suggestions. Thank you!

toumorokoshi committed:
Addressing feedback.

toumorokoshi committed:
Merge branch 'master' into toum/require-field-behavior

toumorokoshi committed:
more fixes

- message to "field"

toumorokoshi committed:
More fixes

- concise wording

shwoodard commented on aip/general/0203.md:
```diff
@@ -16,22 +16,24 @@ behavior (for example, to optimize client library signatures).
-Additionally, APIs **may** use the `OPTIONAL` value to describe none of the
-above. However, it is never mandatory to explicitly describe a field as
-optional.
+- APIs **must** use the `google.api.field_behavior` annotation on every field.
```
nit:

> ...**must** apply the...annotation to every...

"apply" or "include" are *slightly* better words here, imho.

toumorokoshi replied:
ack. I'll use the "apply" term, I think it's a little more direct than include.

shwoodard commented on aip/general/0203.md:
```diff
@@ -16,22 +16,24 @@ behavior (for example, to optimize client library signatures).
-optional.
+- APIs **must** use the `google.api.field_behavior` annotation on every field.
+- If an option described in this AIP accurately describes the behavior of the
+  field, it **must** be annotated on the field.
```
I'm still struggling with this part of the guidance. My concerns are the following:

* Having "if" and **must** confuses the issue.
* "an option described in the AIP" is vague because I don't knew the exact set you are referring too; values are discussed throughout. I would either,
  * explicitly say something like, "The value of the annotation **must** be one of the those described in the`[Vocabulary](#vocabulary)` section of this AIP".
  * or, link to the enum with the values using similar verbiage as that provided above, https://github.com/googleapis/googleapis/blob/master/google/api/field_behavior.proto

The advantage to linking to the proto is that when values are added, this is capture. For example, "Vocabulary" does not cover, `NON_EMPTY_DEFAULT`. If you intend to go with Option 1 above, I would suggest adding a section for `NONE_EMPTY_DEFAULT`. In fact, we should probably do this regardless.

If you do link to the file in googleapis/googleapis, again, I would be explicit as possible, e.g. make sure it is clear that the `FIELD_BEHAVIOR_UNSPECIFIED` value **must not** be used.

toumorokoshi replied:
I definitely see your point. My intention was to force someone to read through
these annotations to see what is relevant and then apply them. 

I like your idea about linking to the proto, I'll do that. How about I just add
a warning here that says "if you add annotations afterward, they could cause
backward-incompatible changes in the clients. Therefore, if it's important to
review the relevant annotations and apply them when the field is introduced.".

or something like that. WDYT?

toumorokoshi replied:
ok PTAL.

toumorokoshi committed:
addressing feedback

- linking to proto as the source-of-truth.
- adding warning on field behavior compatibility.

toumorokoshi committed:
fix minor typo
